### PR TITLE
[REVIEW] new aggregation infrastructure in column reduction

### PR DESCRIPTION
**Pull request by karthikeyann**
Unify aggregation APIs  for reduction https://github.com/rapidsai/cudf/issues/3623

This PR has changes to move to new aggregation infrastructure. These are non-functional changes.

**Files and Changes:**
`include/cudf/detail/aggregation/aggregration.hpp`
- [x]  add in enum Kind
 PRODUCT
 SUMOFSQUARES
 ANY
 ALL 
- [x]  add target_type_impl<> specialization
- [x]  add aggregation_dispatcher() case.
`include/cudf/aggregation.hpp`
- [x]  add make_*_aggregation();
`include/cudf/detail/aggregation/aggregation.cuh`
- [x]  add corresponding_operator<>
`src/cudf/src/aggregation/aggregation.cpp`
- [x]  add make_*_aggregation();
`src/cudf/src/reductions/reductions.cpp`
- [x] replace enums with aggregation::Kind
- [x] create functor to call with aggregation_dispatcher
- [x] Update unit tests

### CODE REVIEW

File: cpp/src/reductions/reductions.cpp [hunk 1]
```diff
@@ -17,70 +17,91 @@
...
     default:
       CUDF_FAIL("Unsupported reduction operator");
+    }
   }
+};
+
+std::unique_ptr<scalar> reduce(
+    column_view const& col, 
+    std::unique_ptr<aggregation> const &agg,
+    data_type output_dtype,
```
Comment (rgsl888prabhu) [inline cpp/src/reductions/reductions.cpp]: Will it be useful if we can derive the output type using the `target_type_t` rather than passing it from API?

Comment (devavret) [reply cpp/src/reductions/reductions.cpp]: Seconded. It will probably greatly reduce compile time

File: cpp/src/reductions/reductions.cpp [hunk 2]
```diff
@@ -17,70 +17,91 @@
...
+                          rmm::mr::device_memory_resource *mr,
+                          cudaStream_t stream)
+      : col(col), output_dtype(output_dtype), mr(mr), stream(stream) {}
 
-  switch (op) {
-    case reduction_op::SUM:
-      result = reduction::sum(col, output_dtype, mr, stream);
+  template <aggregation::Kind k>
+  std::unique_ptr<scalar> operator()(std::unique_ptr<aggregation> const &agg) {
+    switch (k) {
```
Comment (devavret) [inline cpp/src/reductions/reductions.cpp]: There is also a new aggregation dispatcher added as part of the infra. Would require you to restructure the `reduction::<agg>` as template specializations though.

Comment (karthikeyann) [reply cpp/src/reductions/reductions.cpp]: I am using `dispatch_aggregation`. I am using this switch statement to keep it more readable. The switch statement arg is a template parameter, so essentially each case statement block is specialization code.

File: cpp/src/reductions/simple.cuh [hunk 3]
```diff
@@ -68,68 +68,40 @@ std::unique_ptr<scalar> simple_reduction(column_view const& col,
...
+  template <typename ElementType>
     static constexpr bool is_supported_v()
     {
-      // disable only for string ElementType except for operators min, max
-      return  !( std::is_same<ElementType, cudf::string_view>::value &&
-              !( std::is_same<Op, cudf::experimental::reduction::op::min>::value ||
-                 std::is_same<Op, cudf::experimental::reduction::op::max>::value ));
+      using ResultType = cudf::experimental::detail::target_type_t<ElementType, k>;
+      // allow only for convertible types
+      // disable for string ElementType except for operators min, max
```
Comment (rgsl888prabhu) [inline cpp/src/reductions/simple.cuh]: How min and max is enabled for string_view here, `!(std::is_same<ElementType, cudf::string_view>::value` always be false for string_view and nothing will be supported.

File: cpp/src/reductions/reductions.cpp (hunk 1 above)
Comment (jrhemstad) [reply cpp/src/reductions/reductions.cpp]: Hold on here. For series level reductions we specifically wanted the option to be able to specify the desired return type.

Comment (karthikeyann) [reply cpp/src/reductions/reductions.cpp]: That's right. In python pandas, dtype can be mentioned as kwargs. 
Last commit is reverted.

**Comment by jrhemstad** [comment]
rerun tests

**Comment by jrhemstad** [comment]
@karthikeyann looks like for some reason the UTILITIES_TESTS are failing in CI. Are you able to repro locally?

**Comment by rgsl888prabhu** [comment]
> @karthikeyann looks like for some reason the UTILITIES_TESTS are failing in CI. Are you able to repro locally?

This solution is this PR https://github.com/rapidsai/cudf/pull/4007

File: cpp/src/reductions/reductions.cpp (hunk 1 above)
Comment (harrism) [reply cpp/src/reductions/reductions.cpp]: This is essential for avoiding overflow when summing small types, e.g. 100M int8s...

**Comment by jrhemstad** [comment]
rerun tests

File: cpp/src/reductions/reductions.cpp (hunk 1 above)
Comment (devavret) [reply cpp/src/reductions/reductions.cpp]: @harrism , actually, `target_type_t` does take care of this by using `int64_t` for all integer reductions.

**Comment by karthikeyann** [comment]
> @karthikeyann looks like for some reason the UTILITIES_TESTS are failing in CI. Are you able to repro locally?

After merge, this issue didn't come locally.

**Comment by karthikeyann** [comment]
rerun tests

**Comment by rgsl888prabhu** [comment]
> > @karthikeyann looks like for some reason the UTILITIES_TESTS are failing in CI. Are you able to repro locally?
> 
> After merge, this issue didn't come locally.

Normally it happens once in a while depending on whether there is some residual values left in the memory.

File: cpp/src/reductions/reductions.cpp (hunk 1 above)
Comment (karthikeyann) [reply cpp/src/reductions/reductions.cpp]: @devavret For all int, it is always `int64_t`. To give a counter example: For `int64_t` column and with python kwarg `dtype=double`, overflow won't happen in this case. if `target_type_t` is used, overflow could happen.

Comment (devavret) [reply cpp/src/reductions/reductions.cpp]: @karthikeyann , just clarifying: I backtrack on my original request to use `target_type_t`. I just mentioned this as an fyi for @harrism.

**Comment by harrism** [comment]
@rgsl888prabhu are there remaining unaddressed comments from your review? Otherwise please approve so we can merge.

**Pull request merged**